packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -471,6 +499,23 @@ class ConnectionStateHandler implements IConnectionStateHandler {
 		this.setConnectionState(ConnectionState.Disconnected, reason, error);
 	}
 
+	public switchFromEstablishingConnectionToDisconnect(reason: string) {
+		if (this._connectionState === ConnectionState.EstablishingConnection) {
```

Comment:
I'd think (based on name of the API) that when it's called, this._connectionState should always be ConnectionState.EstablishingConnection, no? I.e. is assert more appropriate than check? Maybe there is some race condition the forces it to be a check, I do not know, but it looks weird.

Reply:
So the logic is anyone can call disconnect on container or container gets closed at anytime and at that time we could be in any state. In case of CatchingUp/Connected, we will move to Disconnected state through disconnect event/disconnectFromDeltaStream logic. If we are in Establishing Connection state however, we need to explicitly move to Disconnected state and that is what this api intends to do. So we can not add an assert here.

Reply:
But is it called only when we are in "Establishing Connection" state? Or in both cases?
Can you please add comments through the code to ensure it's clear about full state matrix?

That said, handling for disconnect() & switchFromEstablishingConnectionToDisconnect() is exactly the same. Arguments are too (there is optional error arg on disconnect).

The part that worries me - it's so much easier to think about state transitions when APIs are either in the form of "move to state X" or in the form of "action X happened, react to it". Here, it feels like it's a mix of both, and that means it will be hard to understand full picture oevr time / we will make mistakes. See my other comment first, though.

Reply:
Made  changes. Please take a look.